pyadjoint/optimization/tao_solver.py

```diff
@@ -115,34 +113,41 @@ def from_petsc(self, y, X):
 
         if y_a.shape != (self.n,):
             raise ValueError("Invalid shape")
-        if len(X) != len(self.indices):
+        if len(X) != len(self._index_concatenated_vecs):
             raise ValueError("Invalid length")
-
-        for (i0, i1), x in zip(self.indices, X):
-            _, x_i1 = x._ad_assign_numpy(x, y_a, offset=i0)
-            if i1 != x_i1:
-                raise ValueError("Invalid index")
+        for i, x in enumerate(X):
+            x._ad_petsc_vec().setArray(y_a[self._index_concatenated_vecs[i]])
```

Comment:
`VecGetSubVector` and `VecRestoreSubVector` can be used here to avoid using Numpy.